src/main/scala/nl.knaw.dans.easy.deposit/docs/DDM.scala

```diff
@@ -100,62 +91,55 @@ object DDM extends SchemedXml with DebugEnhancedLogging {
-      case Relation(_, Some(url: String), None) => <label href={ url }>{ url }</label>
-      case Relation(_, None, Some(title: String)) => <label xml:lang={ lang }>{ title }</label>
-      case RelatedIdentifier(Some("id-type:DOI"), Some(value), _) => <label scheme="id-type:DOI" href={ "https://doi.org/" + value }>{ value }</label>
-      case RelatedIdentifier(Some("id-type:URN"), Some(value), _) => <label scheme="id-type:URN" href={ "http://persistent-identifier.nl/" + value }>{ value }</label>
-      case RelatedIdentifier(Some(scheme @ ("id-type:URI" | "id-type:URL")), Some(value), _) => <label scheme={ scheme } href={ value }>{ value }</label>
-      case RelatedIdentifier(scheme, Some(value), _) => <label xsi:type={ scheme.nonBlankOrNull }>{ value }</label>
-      // should not get at the next because of withNonEmpty
-      case _ => throw new IllegalArgumentException("invalid relation " + JsonUtil.toJson(relation))
-    }
-    }.withLabel(relation match { // replace the namespace in case of an href=URL attribute
-    case RelatedIdentifier(_, _, None) => throw new IllegalArgumentException("missing qualifier: RelatedIdentifier" + JsonUtil.toJson(relation))
-    case RelatedIdentifier(Some("id-type:URI" | "id-type:URL" | "id-type:URN" | "id-type:DOI"), _, Some(qualifier)) => qualifier.toString.replace("dcterms", "ddm")
-    case RelatedIdentifier(_, _, Some(qualifier)) => qualifier.toString
-    case Relation(None, _, _) => throw new IllegalArgumentException("missing qualifier: Relation" + JsonUtil.toJson(relation))
-    case Relation(Some(qualifier), Some(_), _) => qualifier.toString.replace("dcterms", "ddm")
-    case Relation(Some(qualifier), _, _) => qualifier.toString
-    // should not get at the next because of withNonEmpty
-    case _ => throw new IllegalArgumentException("invalid relation" + JsonUtil.toJson(relation))
-  })
+  private def basicContent(relation: RelationType, lang: String): Elem = {
+    val cleanRelation = relation.withCleanOptions
+    (cleanRelation match {
+      case Relation(_, Some(url: String), None) =>
+        <label href={ relation.urlOrNull }>{ url }</label>
+      case rel: RelatedIdentifier if rel.url.isEmpty =>
+        <label xsi:type={ rel.schemeOrNull }>{ rel.valueOrThrow }</label>
+      case rel: RelatedIdentifier =>
+        <label scheme={ rel.schemeOrNull } href={ rel.urlOrNull }>{ rel.valueOrThrow }</label>
+      case rel: Relation => // Relation(_,None,None) is skipped so we do have a title (via value) and therefore a language
+        <label xml:lang={ lang } href={ rel.urlOrNull }>{ rel.valueOrThrow }</label>
+    }).withLabel(qualifier(cleanRelation))
+  }
 
-  private def details(source: SchemedKeyValue, label: String, lang: String): Elem = {
-    (label, source) match {
+  private def qualifier(cleanRelation: RelationType) = {
+    if (cleanRelation.url.isEmpty) cleanRelation.qualifierOrThrow
```

Comment:
this `isEmpty` is confusing me. It might be from `Option.isEmpty` or `String.isEmpty`.
how about:

```scala
cleanRelation.url
  .map(_ => cleanRelation.qualifierOrThrow.replace("dcterms", "ddm"))
  .getOrElse(cleanRelation.qualifierOrThrow)
```